atomics/include/desul/atomics/Lock_Free_Fetch_Op.hpp

```diff
@@ -21,40 +21,42 @@ SPDX-License-Identifier: (BSD-3-Clause)
-                                                                                     \
-    return reinterpret_cast<T&>(oldval);                                             \
+#define DESUL_IMPL_ATOMIC_FETCH_OPER(ANNOTATION, HOST_OR_DEVICE)                      \
+  template <class Oper,                                                               \
+            class T,                                                                  \
+            class MemoryOrder,                                                        \
+            class MemoryScope,                                                        \
+            std::enable_if_t<HOST_OR_DEVICE##_atomic_always_lock_free<T>(), int> = 0> \
+  ANNOTATION T HOST_OR_DEVICE##_atomic_fetch_oper(                                    \
+      const Oper& op,                                                                 \
+      T* const dest,                                                                  \
+      dont_deduce_this_parameter_t<const T> val,                                      \
+      MemoryOrder order,                                                              \
+      MemoryScope scope) {                                                            \
+    using cas_t = atomic_compare_exchange_t<T>;                                       \
+    cas_t oldval = reinterpret_cast<cas_t&>(*dest);                                   \
+    cas_t assume = oldval;                                                            \
+                                                                                      \
+    do {                                                                              \
+      if (check_early_exit(op, reinterpret_cast<T&>(oldval), val))                    \
+        return reinterpret_cast<T&>(oldval);                                          \
+      assume = oldval;                                                                \
+      T newval = op.apply(reinterpret_cast<T&>(assume), val);                         \
+      oldval =                                                                        \
+          HOST_OR_DEVICE##_atomic_compare_exchange(reinterpret_cast<cas_t*>(dest),    \
+                                                   assume,                            \
+                                                   reinterpret_cast<cas_t&>(newval),  \
+                                                   order,                             \
+                                                   scope);                            \
+    } while (assume != oldval);                                                       \
+                                                                                      \
+    return reinterpret_cast<T&>(oldval);                                              \
   }
 
 DESUL_IMPL_ATOMIC_FETCH_OPER(DESUL_IMPL_HOST_FUNCTION, host)
+#if defined(DESUL_HAVE_DEVICE_ATOMICS)
```

Comment:
I do not love that new macro.
What toolchain were complaining about this not being defined?
I did not get any issue w/o guarding on my local setup (AppleClang on OSX)